src/es6-module-utils.ts

```diff
@@ -0,0 +1,444 @@
+    // For each document loaded from the analyzer, we build a map of the
+    // original specifiers to the resolved URLs since we want to use analyzer
+    // resolutions for such things as bare module specifiers.
+    const jsImportResolvedUrls =
+        new Map<ResolvedUrl, Map<string, ResolvedUrl>>();
+    const rollupBundle = await rollup({
+      input,
+      external,
+      onwarn: (warning: string) => {},
+      treeshake: false,
+      plugins: [
+        {
+          name: 'analyzerPlugin',
+          resolveId: (importee: string, importer?: string) => {
+            if (importee === input) {
+              return input;
+            }
+            if (importer) {
+              if (jsImportResolvedUrls.has(importer as ResolvedUrl)) {
+                const resolutions =
+                    jsImportResolvedUrls.get(importer as ResolvedUrl)!;
+                if (resolutions.has(importee)) {
+                  return resolutions.get(importee);
+                }
+              }
+              return this.bundler.analyzer.urlResolver.resolve(
+                         importer === input ? url : importer as ResolvedUrl,
+                         importee as FileRelativeUrl)! as string;
+            }
+            return this.bundler.analyzer.urlResolver.resolve(
+                       importee as PackageRelativeUrl)! as string;
+          },
+          load: (id: ResolvedUrl) => {
+            if (id === input) {
+              return code;
+            }
+            if (this.bundle.bundle.files.has(id)) {
+              const document = getAnalysisDocument(analysis, id);
+              if (!jsImportResolvedUrls.has(id)) {
+                const jsImports = document.getFeatures({
+                  kind: 'js-import',
+                  imported: false,
+                  externalPackages: true,
+                  excludeBackreferences: true,
+                });
+                const resolutions = new Map<string, ResolvedUrl>();
+                jsImportResolvedUrls.set(id, resolutions);
+                for (const jsImport of jsImports) {
+                  const source = jsImport.astNode && jsImport.astNode.source &&
+                      jsImport.astNode.source.value;
+                  if (source) {
+                    resolutions.set(source, jsImport.document.url);
+                  }
+                }
+              }
+              return document.parsedDocument.contents;
+            }
+          },
+        },
+      ],
+      experimentalDynamicImport: true,
+    });
+    const {code: rolledUpCode} = await rollupBundle.generate({
+      format: 'es',
+      freeze: false,
+    });
+    const babelFile = parseModuleFile(url, rolledUpCode);
+    this._rewriteImportStatements(url, babelFile);
+    this._deduplicateImportStatements(babelFile);
+    const {code: rewrittenCode} = serialize(babelFile);
+    return {code: rewrittenCode, map: undefined};
+  }
+
+  private _deduplicateImportStatements(node: babel.Node) {
```

Comment:
maybe add little examples to each of these transformation functions to show the input/output?

Reply:
Added in commit decca6d9131f30c1ff28bd25742464fd6b2cedd9